### DISCUSSION TITLE
RuntimeError: expected scalar type BFloat16 but found Float on Raspberry Pi install. Attempting training

### DISCUSSION THREAD

**Discussion by 97Cweb** [coqui-ai/TTS#2408] [General Q&A]
Good evening all
I am trying to get training to work on a raspberry pi as it is the only linux based system I have access to (I am typing on Windows currently) and I have managed to get STT installed, and the basic TTS to work as well. I am running into an issue trying to get the train.py file to run, it spits out the following cascading error:

`
/usr/bin/python3.9 /home/sox/Documents/voice-assistant/TTS Config/train_vits.py --precision full 
 > Setting up Audio Processor...
 | > sample_rate:22050
 | > resample:False
 | > num_mels:80
 | > log_func:np.log10
 | > min_level_db:0
 | > frame_shift_ms:None
 | > frame_length_ms:None
 | > ref_level_db:None
 | > fft_size:1024
 | > power:None
 | > preemphasis:0.0
 | > griffin_lim_iters:None
 | > signal_norm:None
 | > symmetric_norm:None
 | > mel_fmin:0
 | > mel_fmax:None
 | > pitch_fmin:None
 | > pitch_fmax:None
 | > spec_gain:20.0
 | > stft_pad_mode:reflect
 | > max_norm:1.0
 | > clip_norm:True
 | > do_trim_silence:False
 | > trim_db:60
 | > do_sound_norm:False
 | > do_amp_to_db_linear:True
 | > do_amp_to_db_mel:True
 | > do_rms_norm:False
 | > db_level:None
 | > stats_path:None
 | > base:10
 | > hop_length:256
 | > win_length:1024
 | > Found 88 files in /home/sox/Documents/voice-assistant/TTS Config/sox-dataset
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
 > Training Environment:
 | > Num. of CPUs: 4
 | > Num. of Torch Threads: 4
 | > Torch seed: 54321
 | > Torch CUDNN: True
 | > Torch CUDNN deterministic: False
 | > Torch CUDNN benchmark: False
 > Start Tensorboard: tensorboard --logdir=vits_ljspeech-March-12-2023_07+59PM-0000000

 > Model has 83059180 parameters

 > EPOCH: 0/100000
 --> vits_ljspeech-March-12-2023_07+59PM-0000000

 > TRAINING (2023-03-12 19:59:21) 

> DataLoader initialization
| > Tokenizer:
	| > add_blank: True
	| > use_eos_bos: False
	| > use_phonemes: True
	| > phonemizer:
		| > phoneme language: en-us
		| > phoneme backend: espeak
| > Number of instances : 83
 | > Preprocessing samples
 | > Max text length: 188
 | > Min text length: 1
 | > Avg text length: 43.34939759036145
 | 
 | > Max audio length: 306297.0
 | > Min audio length: 11488.0
 | > Avg audio length: 55226.56626506024
 | > Num. instances discarded samples: 0
 | > Batch group size: 48.
/home/sox/.local/lib/python3.9/site-packages/torch/functional.py:632: UserWarning: stft will soon require the return_complex parameter be given for real inputs, and will further require that return_complex=True in a future PyTorch release. (Triggered internally at /root/pytorch/aten/src/ATen/native/SpectralOps.cpp:801.)
  return _VF.stft(input, n_fft, hop_length, win_length, window,  # type: ignore[attr-defined]
 ! Run is removed from vits_ljspeech-March-12-2023_07+59PM-0000000
Traceback (most recent call last):
  File "/home/sox/.local/lib/python3.9/site-packages/trainer/trainer.py", line 1591, in fit
    self._fit()
  File "/home/sox/.local/lib/python3.9/site-packages/trainer/trainer.py", line 1544, in _fit
    self.train_epoch()
  File "/home/sox/.local/lib/python3.9/site-packages/trainer/trainer.py", line 1309, in train_epoch
    _, _ = self.train_step(batch, batch_num_steps, cur_step, loader_start_time)
  File "/home/sox/.local/lib/python3.9/site-packages/trainer/trainer.py", line 1162, in train_step
    outputs, loss_dict_new, step_time = self._optimize(
  File "/home/sox/.local/lib/python3.9/site-packages/trainer/trainer.py", line 1023, in _optimize
    outputs, loss_dict = self._model_train_step(batch, model, criterion, optimizer_idx=optimizer_idx)
  File "/home/sox/.local/lib/python3.9/site-packages/trainer/trainer.py", line 970, in _model_train_step
    return model.train_step(*input_args)
  File "/home/sox/Documents/voice-assistant/TTS/TTS/tts/models/vits.py", line 1252, in train_step
    outputs = self.forward(
  File "/home/sox/Documents/voice-assistant/TTS/TTS/tts/models/vits.py", line 1024, in forward
    outputs, attn = self.forward_mas(outputs, z_p, m_p, logs_p, x, x_mask, y_mask, g=g, lang_emb=lang_emb)
  File "/home/sox/Documents/voice-assistant/TTS/TTS/tts/models/vits.py", line 921, in forward_mas
    loss_duration = self.duration_predictor(
  File "/home/sox/.local/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/sox/Documents/voice-assistant/TTS/TTS/tts/layers/vits/stochastic_duration_predictor.py", line 238, in forward
    x = self.convs(x, x_mask)
  File "/home/sox/.local/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/sox/Documents/voice-assistant/TTS/TTS/tts/layers/vits/stochastic_duration_predictor.py", line 56, in forward
    y = self.norms_1[i](y)
  File "/home/sox/.local/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/sox/Documents/voice-assistant/TTS/TTS/tts/layers/generic/normalization.py", line 52, in forward
    x = torch.nn.functional.layer_norm(x, (self.channels,), self.gamma, self.beta, self.eps)
  File "/home/sox/.local/lib/python3.9/site-packages/torch/nn/functional.py", line 2515, in layer_norm
    return torch.layer_norm(input, normalized_shape, weight, bias, eps, torch.backends.cudnn.enabled)
RuntimeError: expected scalar type BFloat16 but found Float
Process finished with exit code 1

As can be seen with the final RuntiumeError, tehre seems to be an issue with BFloat16 and Raspberry Pi. I do not know enough about how to make this work from here to fix it. I searched online and it said to try using the command --precision full but that did not help either (you can see it in the top of the crash report above)

Any help is greatly appreciated. I know it will take time to train, but I don't have anything better, and it can run for a while on the Pi alone

**Comment by alexbanta4** [comment]
I ran into a similar issue, and it turned out to be related to torch autolog, which runs into issues with CPUs: 'For CPU, only lower precision floating point datatype of torch.bfloat16 is supported for now.' I was not running on a RaspberryPi and fixed the issue by switching to a GPU, but if you cannot do something analogous to that, [this post](https://stackoverflow.com/questions/75905014/while-running-stable-diffusion-and-torch-on-cpu-runtimeerror-expected-scalar-ty) may be helpful

**Reply by 97Cweb** [reply]
Thank you for that post. From what I read in it, it means that the cpu does not support Float16, and there is no way around it. I'll have to generate on a different computer and push the final file over.
